pallets/parachain-staking/src/lib.rs

```diff
@@ -2266,27 +2824,43 @@ pub mod pallet {
 					amount >= T::MinDelegatorStk::get(),
 					Error::<T>::DelegatorBondBelowMin
 				);
-				ensure!(!Self::is_candidate(&acc), Error::<T>::CandidateExists);
-				Delegator::new(acc.clone(), collator.clone(), amount)
+				ensure!(!Self::is_candidate(&delegator), Error::<T>::CandidateExists);
+				Delegator::new(delegator.clone(), candidate.clone(), amount)
 			};
-			let mut state = <CandidateState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+			let mut state = <CandidateInfo<T>>::get(&candidate).ok_or(Error::<T>::CandidateDNE)?;
 			ensure!(
-				candidate_delegation_count >= state.delegators.0.len() as u32,
+				candidate_delegation_count >= state.delegation_count,
 				Error::<T>::TooLowCandidateDelegationCountToDelegate
 			);
-			let delegator_position = state.add_delegation::<T>(acc.clone(), amount)?;
-			T::Currency::reserve(&acc, amount)?;
+			let (delegator_position, less_total_staked) = state.add_delegation::<T>(
```

Comment:
From my little knowledge of the code, I have the impression that this is modifying the state before we actually check whether the delegator can bond (reserve) the specified amount. If `T::Currency::reserve(&delegator, amount)?;` fails would  not we have already modified the storage in `state.add_delegation::<T>`

Reply:
Good point, it was writing to the top and bottom delegations before doing the reserve check. So I moved a `can_reserve` check to the top of the function body to ensure the caller can reserve before modifying any state.